audiorecorder.1
.TH audiorecorder 1 "github.com/amiaopensource/audiorecorder" "2019-05-21" "AMIA Open Source"
.\" Turn off justification for nroff.
.if n .ad l
.\" Turn off hyphenation.
.nh
.SH NAME
audiorecorder - Tool for calibration and recording of analog audio sources
.SH SYNOPSIS
\fBaudiorecorder\fR [\fB-e\fR | \fB-p\fR | \fB-m\fR | \fB-b\fR | \fB-h\fR]
.SH OPTIONS
.TP
.B -e
edit config
.TP
.B -p
passthrough mode
.TP
.B -m
edit metadata to be used in BWF creation
.TP
.B -b
embed BEXT chunk (BWF metadata) in target file
.TP
.B -h
display a help message
.SH LICENSE
.TP
The audiorecorder script is covered under a BSD 3-clause license.
.TP
All documentation for audiorecorder is licensed under a Creative Commons Attribution 4.0 International License.


history.md
# Development History

**2018-08-30:** audiorecorder2 added to homebrew install formula. Can now be installed with option `--with-audiorecorder2`

**2018-03-15:** In development version 2 (in linux) tested with twenty minute recording at 24 bit/96 kHz stereo recording. File analyzed for interstitial errors using Wavelab Elements 9.5 global analysis with none detected.

**2017-12-26:** Version 0.1.08 released. Mainly typo fixes.

**2017-12-14:** Version 0.1.07 released. Fix bug. Stricter man page syntax.

**2017-10-18:** Version 0.1.06 released. Improved help message and man page.

**2017-10-01:** Version 0.1.05 released. Deletes BWF Metaedit from the package and uses its homebrew formula instead.

**2017-08-22:** Version 0.1.04 released. Mainly doc fixes.

**2017-06-05:** Version 0.1.03 released. Linux install instructions, man page and Code of Conduct have been added.

**2017-04-25:** Version 0.1.02 released. Includes font support for filters in Linux. (Tested in Ubuntu 16.04).

**2017-04-11:** Version 0.1.01 released. Contains some basic interface workarounds to decrease reliance on Pashua for user input. This makes audiorecorder able to run in a linux environment (further reliability testing for non macOS systems is underway).

**2017-03-11:** Major redesign of post-digitization functions and GUI. Audiorecorder now supports trimming of start and end of files via manual specification as well as auto-trim of silence at the start of files. Preview and post-digitization GUI now incorporate waveforms of the digitized audio to aid in trimming. All testing has been negative for dropped samples.

**2017-03-04:** For the new release changes have been made to streamline the way audiorecorder handles input channels. This will prevent it from being overwhelmed by A/D converters that output many empty tracks along with the desired tracks (such as the Apogee Symphony). In two 55 minute tests and one 30 minute no drops or problems were detected via Wavelab analysis. Due to more efficient management of data the current buffering now has increased the latency for mono captures. A next step will be to see how much buffering can be safely lowered.

**2017-02-24:** Changes appear to have been successful for adapting audiorecorder to macOS 10.12.03. In testing it was discovered that audiorecorder has issues with A/D converters that have a large number of output tracks (such as sixteen as opposed to two). Current audiorecorder head appears to be stable for two track A/D converters and testing is underway to expand its ability to deal with larger multi-track converters.

**2017-02-23:** Due to some issues that were reported with macOS 10.12.03, some changes to buffering have been made and are currently being tested.

**2017-02-15:** Ability to select which channel to record has been added to the GUI. This option was tested with a 25 minute ingest on 2013 Macbook Air with no dropped samples detected audibly or vie Wavelab global analysis. Some glitches with file trimming were resolved.

**2017-01-20:** Initial post digitization functions have been included via a basic GUI. These include ability to preview file with and without silence trimming, and an option to create a silence trimmed version of file. The spectrograph has also been changed to scroll for easier viewability. Further testing has not detected any dropped samples either audibly or via Wavelab global analysis. 

**2016-12-22:** Current build was tested with a 40 minute transfer on the 2013 Macbook Air with no dropped samples detected either audibly or via Wavelab's global analysis tool. Testing continues...


audiorecorder.sh
#!/usr/bin/env bash
#Cascadia Now!
config="$HOME/.$(basename "${0}").conf"
bwf_config="$HOME/.$(basename "${0}")_BWF.conf"
touch "${config}"
touch "${bwf_config}"
. "$config"
. "${bwf_config}"



OPTIND=1
while getopts "epm" opt ; do
    case "${opt}" in
        e) runtype="edit";;
        p) runtype="passthrough";;
        m) runtype="metadata";;
        *)
    esac
done

_lookup_devices(){
    OLDIFS="$IFS"
    device_list=("$(ffmpeg -hide_banner -f avfoundation -list_devices true -i "" -f null /dev/null 2>&1 | cut -d ' ' -f6- | sed -e                  '1,/AVFoundation audio devices:/d' | grep -v '^$'| tr '\n' ',' )")
    IFS=',' read -r -a DEVICES <<< "$device_list"
    IFS="$OLDIFS"
}
_lookup_sample_rate(){
    case "${1}" in
        "44.1 kHz")
            SAMPLE_RATE_NUMERIC="44100"
            ;;
        "48 kHz")
            SAMPLE_RATE_NUMERIC="48000"
            ;;
        "96 kHz")
            SAMPLE_RATE_NUMERIC="96000"
            ;;
    esac
}

_lookup_bit_depth(){
    case "${1}" in
        "16 bit")
            CODEC="pcm_s16le"
            ;;
        "24 bit")
            CODEC="pcm_s24le"
            ;;
    esac
}
_get_ITEM_ID(){
    echo "Please Input Item ID or type Q to exit"
    read ITEM_ID
    if [ "${ITEM_ID}" = "Q" ] || [ "${ITEM_ID}" = "q" ] ; then
        echo "Goodbye!"
        exit 0
    fi
    if [ -f "${output}/${ITEM_ID}.wav" ] ; then
        echo "A file with that name already exists!"
        _get_ITEM_ID
    fi
}

_metadata_gui(){
    gui_conf="
    # Set transparency: 0 is transparent, 1 is opaque
    *.transparency=0.95
    # Set window title
    *.title = Edit Metadata Values
    # intro text
    intro.width = 300
    intro.type = text
    intro.text = Hello
    #Originator
    originator.type = textfield
    originator.label = Enter Originator
    originator.default = "${originator}"
    #Coding History
    coding_history.type = textbox
    coding_history.label = Enter Coding History
    coding_history.default = "${coding_history}"
    cb.type = cancelbutton
    cb.label = Cancel
    "

    pashua_configfile=`/usr/bin/mktemp /tmp/pashua_XXXXXXXXX`
    echo "${gui_conf}" > "${pashua_configfile}"
    pashua_run
    rm "${pashua_configfile}"
}

_master_gui(){
    _lookup_devices    
    gui_conf="
    # Set transparency: 0 is transparent, 1 is opaque
    *.transparency=0.95
    # Set window title
    *.title = Welcome to Audio Recorder!
    # intro text
    intro.width = 300
    intro.type = text
    intro.text = Hello
    #Output Location
    output.type = openbrowser
    output.label = Output Location
    output.default = "${output}"
    #Capture Device
    device.type = popup
    device.label = Select Audio Capture Device
    device.option = ${DEVICES[0]}
    device.option = ${DEVICES[1]}
    device.option = ${DEVICES[2]}
    device.option = ${DEVICES[3]}
    device.default = "${device}"
    #Sample Rate
    sample_rate.type = radiobutton
    sample_rate.label = Select Sample Rate
    sample_rate.option = 44.1 kHz
    sample_rate.option = 48 kHz
    sample_rate.option = 96 kHz
    sample_rate.default = "${sample_rate}"
    #Bit Depth
    bit_depth.type = radiobutton
    bit_depth.label = Select Bit Depth
    bit_depth.option = 16 bit
    bit_depth.option = 24 bit
    bit_depth.default = "${bit_depth}"
    #Cancel Button
    cb.type = cancelbutton
    cb.label = Cancel
    "

    pashua_configfile=`/usr/bin/mktemp /tmp/pashua_XXXXXXXXX`
    echo "${gui_conf}" > "${pashua_configfile}"
    pashua_run
    rm "${pashua_configfile}"
}
pashua_run() {
    # Wrapper function for interfacing to Pashua. Written by Carsten
    # Bluem <carsten@bluem.net> in 10/2003, modified in 12/2003 (including
    # a code snippet contributed by Tor Sigurdsson), 08/2004 and 12/2004.
    # Write config file

    # Find Pashua binary. We do search both . and dirname "$0"
    # , as in a doubleclickable application, cwd is /
    # BTW, all these quotes below are necessary to handle paths
    # containing spaces.
    bundlepath="Pashua.app/Contents/MacOS/Pashua"
    mypath=$(dirname "${0}")
    for searchpath in "$mypath/Pashua" "$mypath/$bundlepath" "./$bundlepath" \
                      "/Applications/$bundlepath" "$HOME/Applications/$bundlepath"
    do
        if [ -f "$searchpath" -a -x "$searchpath" ] ; then
            pashuapath=$searchpath
            break
        fi
    done
    if [ ! "$pashuapath" ] ; then
        echo "Error: Pashua is not found."
                break 2
            fi
        encoding=""
        # Get result
        result=`"$pashuapath" $encoding $pashua_configfile | sed 's/ /;;;/g'`

        # Parse result
        for line in $result ; do
            key=`echo $line | sed 's/^\([^=]*\)=.*$/\1/'`
            value=`echo $line | sed 's/^[^=]*=\(.*\)$/\1/' | sed 's/;;;/ /g'`
            varname=$key
            varvalue="$value"
            eval $varname='$varvalue'
        done
} # pashua_run()

if [ "${runtype}" = "edit" ] ; then
_master_gui

{
    echo "output=\"${output}\""
    echo "device=\"${device}\""
    echo "sample_rate=\"${sample_rate}\""
    echo "bit_depth=\"${bit_depth}\""
} > "${config}"
fi

if [ "${runtype}" = "metadata" ] ; then
_metadata_gui

{
    echo "originator=\"${originator}\""
    echo "coding_history=\"${coding_history}\""
} > "${bwf_config}"
fi

if [ -n ${DEVICE_NUMBER} ] ; then
    _lookup_sample_rate "${sample_rate}"
else 
    echo "No Sample Rate Specified.  Setting to 96kH."
    SAMPLE_RATE_NUMERIC="96000"
fi
if [ -n "${bit_depth}" ] ; then
    _lookup_bit_depth  "${bit_depth}"
else 
    echo "No Bit Depth Specified.  Setting to 24."
    CODEC="pcm_s24le"
fi
if [ -n "${device}" ] ; then
    DEVICE_NUMBER=$(echo "${device}" | cut -c 2)
else 
    echo "No Device Specified.  Attempting to Guess Device."
    DEVICE_NUMBER="0"
fi
if [ -z "${output}" ] ; then
    echo "No Output Location Specified.  Setting Output to Desktop"
    output="~/Desktop"
fi


FILTER_CHAIN="asplit=6[out1][a][b][c][d][e],\
[e]showvolume=w=700:c=0xff0000:r=30[e1],\
[a]showfreqs=mode=bar:cmode=separate:size=300x300:colors=magenta|yellow[a1],\
[a1]drawbox=12:0:3:300:white@0.2[a2],[a2]drawbox=66:0:3:300:white@0.2[a3],[a3]drawbox=135:0:3:300:white@0.2[a4],[a4]drawbox=202:0:3:300:white@0.2[a5],[a5]drawbox=271:0:3:300:white@0.2[aa],\
[b]avectorscope=s=300x300:r=30:zoom=5[b1],\
[b1]drawgrid=x=150:y=150:c=white@0.3[bb],\
[c]showspectrum=s=400x600:mode=combined:color=rainbow:scale=lin:saturation=4[cc],\
[d]astats=metadata=1:reset=1,adrawgraph=lavfi.astats.Overall.Peak_level:max=0:min=-30.0:size=700x256:bg=Black[dd],\
[dd]drawbox=0:0:700:42:hotpink@0.2:t=42[ddd],\
[aa][bb]vstack[aabb],[aabb][cc]hstack[aabbcc],[aabbcc][ddd]vstack[aabbccdd],[e1][aabbccdd]vstack[z],\
[z]drawtext=fontfile=/Library/Fonts/Andale Mono.ttf: text='%{pts \\: hms}':x=460: y=50:fontcolor=white:fontsize=30:box=1:boxcolor=0x00000000@1[fps],[fps]fps=fps=30[out0]"


if [ "${runtype}" = "passthrough" ] ; then
    ffmpeg -f avfoundation -i "none:"${DEVICE_NUMBER}"" -f wav -c:a pcm_s16le -ar 44100 - |\
    ffplay -window_title "Audio Record" -f lavfi \
    "amovie='pipe\:0',${FILTER_CHAIN}"
    exit
fi

_get_ITEM_ID

mkfifo PIPE2REC
ffmpeg -f avfoundation -i "none:"${DEVICE_NUMBER}"" -f wav -c:a "${CODEC}" -ar "${SAMPLE_RATE_NUMERIC}" -y PIPE2REC -f wav -c:a pcm_s16le -ar 44100 - |\
ffplay -window_title "Audio Record" -f lavfi \
"amovie='pipe\:0',${FILTER_CHAIN}" | ffmpeg -i PIPE2REC -c copy -rf64 auto "${output}"/"${ITEM_ID}".wav


# Embed metadata in BEXT
echo "Preparing to embed metadata"
echo ""
if
	(("${#ITEM_ID}" > 32));
then
	orig_ref="See description for Identifiers"
else
    orig_ref="${ITEM_ID}"
fi

bwfmetaedit --reject-overwrite --Description="${ITEM_ID}".wav --Originator="${originator}" --OriginatorReference="${orig_ref}" --History="${coding_history}" --IARL="${originator}" --MD5-Embed --OriginationDate=$(date "+%Y-%m-%d") --OriginationTime=$(date "+%H:%M:%S") "${output}"/"${ITEM_ID}".wav
